src/ml/neural_net/combine_iterator.hpp

```diff
@@ -0,0 +1,147 @@
+
+  /** Returns the next value in the sequence. May throw on error. */
+  virtual Output Next() = 0;
+
+  /** Returns a Publisher wrapping this Iterator. */
+  std::shared_ptr<IteratorPublisher<T>> AsPublisher() {
+    return std::make_shared<IteratorPublisher<T>>(this->shared_from_this());
+  }
+};
+
+/**
+ * Concrete Publisher that wraps an Iterator.
+ *
+ * The resulting Publisher is unicast: each iterated value will go only to
+ * whichever Subscriber triggered the iteration.
+ */
+template <typename T>
+class IteratorPublisher : public Publisher<T> {
+ public:
+  using Output = T;
+
+  explicit IteratorPublisher(std::shared_ptr<Iterator<Output>> iterator)
+      : iterator_(std::move(iterator)) {}
+
+  void Receive(std::shared_ptr<Subscriber<Output>> subscriber) override {
+    auto subscription =
+        std::make_shared<IteratorSubscription>(subscriber, iterator_);
+    subscriber->Receive(std::move(subscription));
+  }
+
+ private:
+  // All of the logic lives in the implementation of Subscription, which relies
+  // on the assumption that only one Subscription at a time will access the
+  // shared Iterator.
+  class IteratorSubscription : public Subscription {
+   public:
+    IteratorSubscription(std::shared_ptr<Subscriber<Output>> subscriber,
+                         std::shared_ptr<Iterator<Output>> iterator)
+        : subscriber_(std::move(subscriber)), iterator_(std::move(iterator)) {}
+
+    bool IsActive() const { return subscriber_ != nullptr; }
+
+    void Cancel() override { subscriber_.reset(); }
+
+    void Request(Demand demand) override {
+      // Keep sending signals to the Subscriber until we're cancelled or we
+      // exhaust the demand.
+      while (IsActive() && !demand.IsNone()) {
+        // Invoke the iterator to determine what signal we'll send.
+
+        // Don't assume that Output has a (cheap) default constructor.
+        // TODO: Use an Optional type instead.
+        using OutputStorage =
+            typename std::aligned_storage<sizeof(Output),
+                                          alignof(Output)>::type;
+        OutputStorage value_storage;
+        Output* value = nullptr;     // Track whether we have a value.
+        std::exception_ptr failure;  // Track whether we have a failure.
+        try {
+          if (iterator_->HasNext()) {
+            // Use placement new to initialize our value from the iterator.
+            value = reinterpret_cast<Output*>(&value_storage);
+            new (value) Output(iterator_->Next());
+          }
+        } catch (...) {
+          // On any exception, *value was not initialized, since it was the last
+          // statement in the try block.
+          value = nullptr;
+          failure = std::current_exception();
+        }
+
+        // Send the appropriate signal.
+        if (failure) {
+          // Signal failure and ensure we don't send any more signals.
+          subscriber_->Receive(Completion::Failure(failure));
```

Comment:
I see! Subscriber is deeply attached to Publisher. That restricts one to one relationships between subscribers and publishers, which means only allows singe streaming. Oh, well, I thought it's one-to-many at the very beginning. It's worth mentioning at the beginning.

In short,
1. publisher holds only one subscriber
2. The only one subscriber holds a subscription from the publisher.

If the publisher and the subscriber only run at most one thread for each of them, this will be good.

Again, I think it's better not to attach a subscriber closely attached to the publisher. You can use the `subscription` object to send` promises` to the publisher's task queue (maybe multiple queues for each subscriber so that one copy of data can be sent to many). That can allow one publisher and multiple subscribers running in their own threads and do the test independently.

Reply:
Yes! It turns out that our use case doesn't really need "multicast" publishing, so I haven't worried too much about this.

We do want to introduce multiple threads at some point. Indeed, that's one of the main values of Combine and similar frameworks. These tend to reason about dispatch to other task queues as "scheduling", but for us, it will probably suffice just to worry about hopping threads (not time delays and such).

So in the example that you specify, you would typically set each subscriber to receive elements on a different queue. Something like:
```
iterator->ReceiveOn(thread1)->Subscribe(subscriber1);
iterator->ReceiveOn(thread2)->Subscribe(subscriber2);
```
Effectively the middle "pipe" in this pipeline just switches to the desired thread before sending the value along

Reply:
Yeah, it depends on the original publisher to decide how to distribute (multicast) the data. It can keep track of different queues for the different subscribers. I think this protocol is very flexible. Learned a ton! Thanks for this great work!

Reply:
Yeah, that's something that wasn't obvious to me: the behavior with multiple subscribers is not all standardized. It's up to the publisher to define. This is one reason that Combine and similar frameworks introduce an explicit `multicast` operator, which supports adding multiple subscribers and then having a trigger/barrier that begins pushing values only after all subscribers are ready.